File: R/fit_partial_credit_model_ncats.R
#' Run partial credit model analysis (ncats version) on pre-processed data
#'
#' This function performs Bayesian IRT analysis using the flexible ncats partial credit model
#' on pre-processed data. It compiles the Stan model, runs MCMC sampling, generates convergence
#' diagnostics, and optionally creates detailed diagnostic plots and posterior predictive checks.
#'
#' @param dit data.table. Item metadata table with item types and labels
#' @param dcati data.table. Pre-processed data with observations for analysis. Must include
#'   item_type, y_stan, item_time_id, pid, and oidt columns.
#' @param output_file_prefix Character. Full path prefix for output files (without extension)
#' @param stan_file Character. Path to Stan model file (.stan). Default: partial_credit_model_ncats_v260413.stan
#' @param x_formula Formula. Formula specifying predictors for the design matrix
#' @param chains Integer. Number of MCMC chains to run (default: 2)
#' @param parallel_chains Integer. Number of chains to run in parallel (default: 2)
#' @param threads_per_chain Integer. Number of threads to use per chain (default: 1)
#' @param iter_warmup Integer. Number of warmup iterations per chain (default: 500)
#' @param iter_sampling Integer. Number of sampling iterations per chain (default: 1500)
#' @param seed Integer. Random seed for reproducibility (default: 123)
#' @param show_messages Logical. If TRUE, show all Stan informational messages during sampling (default: FALSE)
#' @param show_exceptions Logical. If TRUE, show detailed Stan exception messages when errors occur (default: FALSE)
#' @param with_core_analyses Logical. If TRUE, generate core probability plots (default: TRUE)
#' @param with_additional_analyses Logical. If TRUE, generate additional diagnostic plots (default: FALSE)
#'
#' @return Invisibly returns the fitted cmdstanr model object
#'
#' @import data.table
#' @import ggplot2
#' @import cmdstanr
#'
#' @export
fit_partial_credit_model_ncats <- function(
  dit,
  dcati,
  output_file_prefix,
  stan_file = here::here("src", "stan", "partial_credit_model_ncats_v260413.stan"),
  x_formula = ~ time - 1,
  chains = 2L,
  parallel_chains = 2L,
  threads_per_chain = 1L,
  iter_warmup = 500L,
  iter_sampling = 1500L,
  seed = 123L,
  show_messages = FALSE,
  show_exceptions = FALSE,
  with_core_analyses = TRUE,
  with_additional_analyses = FALSE
) {
    # Suppress data.table NSE warnings
    item_type <- y_stan <- item_time_id <- pid <- time <- oidt <- variable <-
        ypred <- oid <- in_ppi <- item_label <- time_label <- y_label <- y <-
        n <- total <- p_emp <- group_label_long <- item_label_short <-
        item_label_long <- ess_bulk <- q_lower <- q_upper <- iqr_lower <-
        iqr_upper <- prob <- cat_type_id <- NULL

    require(data.table)
    require(ggplot2)
    require(ggsci)
    require(bayesplot)
    require(cmdstanr)

    # Print configuration
    cat("\n========================================\n")
    cat("Partial Credit Model (ncats) Analysis Configuration\n")
    cat("========================================\n")
    cat("Stan file:", stan_file, "\n")
    cat("Stan include dir:", dirname(stan_file), "\n")
    cat("Data: dit with", nrow(dit), "items, dcati with", nrow(dcati), "observations\n")
    cat("Output prefix:", output_file_prefix, "\n")
    cat("Chains:", chains, "\n")
    cat("Parallel chains:", parallel_chains, "\n")
    cat("Threads per chain:", threads_per_chain, "\n")
    cat("Warmup iterations:", iter_warmup, "\n")
    cat("Sampling iterations:", iter_sampling, "\n")
    cat("Seed:", seed, "\n")
    cat("Core analyses:", with_core_analyses, "\n")
    cat("Additional analyses:", with_additional_analyses, "\n")
    cat("========================================\n\n")

    # Create output directory if it doesn't exist
    dir.create(dirname(output_file_prefix), showWarnings = FALSE, recursive = TRUE)

    # Compile Stan model
    cat("Compiling Stan model...\n")
    pcm_compiled <- cmdstanr::cmdstan_model(
        stan_file,
        include_paths = dirname(stan_file),
        cpp_options = list(stan_threads = TRUE)
    )

    # Prepare data in ncats format
    cat("Preparing Stan data in ncats format...\n")
    
    # Get unique item types (sorted)
    item_types <- sort(unique(dcati$item_type))
    C <- length(item_types)
    
    # Assign cat_type_id based on sorted item types
    dcati_copy <- copy(dcati)
    dcati_copy[, cat_type_id := match(item_type, item_types)]
    
    # Sort data by cat_type_id to ensure proper ordering
    setkey(dcati_copy, cat_type_id, oidt)
    
    # Build stan_data
    stan_data <- list()
    stan_data$C <- C
    stan_data$U <- max(dcati_copy$pid)
    
    # Arrays for each category type
    stan_data$N <- integer(C)
    stan_data$Q <- integer(C)
    stan_data$K <- integer(C)
    
    for (i in 1:C) {
        item_type_i <- item_types[i]
        stan_data$N[i] <- nrow(dcati_copy[item_type == item_type_i])
        stan_data$Q[i] <- max(dcati_copy[item_type == item_type_i, item_time_id])
        stan_data$K[i] <- length(unique(dcati_copy[item_type == item_type_i, y_stan]))
    }
    
    # Total counts
    stan_data$N_total <- sum(stan_data$N)
    stan_data$Q_total <- sum(stan_data$Q)
    
    # Concatenated observation data (data is already sorted by cat_type_id)
    stan_data$y <- dcati_copy$y_stan
    stan_data$unit_of_obs <- dcati_copy$pid
    stan_data$question_of_obs <- dcati_copy$item_time_id
    stan_data$cat_type <- dcati_copy$cat_type_id
    
    # Create design matrix for all observations
    stan_data$X <- model.matrix(
        x_formula,
        data = as.data.frame(dcati_copy)
    )
    stan_data$P <- ncol(stan_data$X)
    
    cat("Design matrix number of predictors: P =", stan_data$P, "\n")
    cat("Design matrix column names:", paste(colnames(stan_data$X), collapse = ", "), "\n")
    cat("Number of category types: C =", C, "\n")
    cat("Category type labels:", paste(item_types, collapse = ", "), "\n")
    cat("N per category:", paste(stan_data$N, collapse = ", "), "\n")
    cat("Q per category:", paste(stan_data$Q, collapse = ", "), "\n")
    cat("K per category:", paste(stan_data$K, collapse = ", "), "\n")

    # Sample from the model
    cat("Running MCMC sampling...\n")
    flush.console()
    pcm_fit <- pcm_compiled$sample(
        data = stan_data,
        seed = seed,
        chains = chains,
        parallel_chains = parallel_chains,
        threads_per_chain = threads_per_chain,
        iter_warmup = iter_warmup,
        iter_sampling = iter_sampling,
        refresh = 500,
        save_warmup = TRUE,
        show_messages = show_messages,
        show_exceptions = show_exceptions
    )
    
    # Remove any trajectories that did not converge
    cat("Checking for divergent transitions and removing non-converged chains...\n")
    tmp <- as.data.table(pcm_fit$draws(variables = "lp__", inc_warmup = FALSE, format = "draws_df"))
    tmp <- tmp[, .(
        mean_lp = mean(lp__),
        sd_lp = sd(lp__),
        n = .N
    ), by = .chain]
    
    threshold <- tmp[which.max(mean_lp), mean_lp - 2 * sd_lp]
    pcm_fit_good_chains <- tmp[mean_lp > threshold, .chain]
    cat("Identified good HMC chains:", paste(pcm_fit_good_chains, collapse = ", "), "\n")

    # Extract chain timing information
    cat("\nExtracting timing information...\n")
    chain_times <- pcm_fit$time()
    timing_data <- data.table(
        chain = pcm_fit_good_chains,
        warmup_minutes = chain_times$chains$warmup[pcm_fit_good_chains] / 60,
        sampling_minutes = chain_times$chains$sampling[pcm_fit_good_chains] / 60,
        total_chain_minutes = chain_times$chains$total[pcm_fit_good_chains] / 60
    )
    
    timing_file <- paste0(output_file_prefix, "_timing.csv")
    write.csv(timing_data, file = timing_file, row.names = FALSE)
    cat("Saved timing information to:", timing_file, "\n")

    # Extract and save draws
    cat("Extracting draws...\n")
    draws_file <- paste0(output_file_prefix, "_draws.rds")
    draws <- pcm_fit$draws(format = "draws_array")
    draws <- draws[, pcm_fit_good_chains, , drop = FALSE]
    saveRDS(draws, file = draws_file)
    cat("Saved draws to:", draws_file, "\n")
    draws <- NULL
    gc()

    # Save output to RDS
    output_file <- paste0(output_file_prefix, "_stan.rds")
    cat("Saving model fit to:", output_file, "\n")
    pcm_fit$save_object(file = output_file)

    # Check convergence and mixing
    cat("Generating convergence diagnostics...\n")
    tmp <- pcm_fit$summary(
        variables = c(
            "latent_factor_unit", "latent_factor_beta",
            "skill_thresholds", "loadings_questions_m1"
        )
    )
    tmp <- as.data.table(tmp)
    tmp <- tmp[order(ess_bulk), ]
    write.csv(
        tmp,
        file = paste0(output_file_prefix, "_convergence_mixing.csv"),
        row.names = FALSE
    )

    # Additional diagnostic analyses (optional)
    if (with_additional_analyses) {
        cat("\nRunning additional diagnostic analyses...\n")

        # Worst parameters with lowest ess_bulk
        worst_var <- tmp$variable[1:9]

        # Make worst trace plot
        cat("Generating trace plots...\n")
        po <- pcm_fit$draws(
            variables = c("lp__", worst_var),
            inc_warmup = TRUE,
            format = "draws_array"
        )
        po <- po[, pcm_fit_good_chains, , drop = FALSE]
        
        lp_range <- range(po[(iter_warmup + 1):(iter_warmup + iter_sampling), , "lp__"])
        lp_ylim <- c(lp_range[1] - 0.05 * diff(lp_range), 
                     lp_range[2] + 0.05 * diff(lp_range))
        
        p_lp <- bayesplot:::mcmc_trace(po[,, "lp__", drop = FALSE],
            pars = "lp__",
            n_warmup = iter_warmup
        ) + 
        coord_cartesian(ylim = lp_ylim) +
        theme_bw()
        
        p_other <- bayesplot:::mcmc_trace(po[,, worst_var, drop = FALSE],
            pars = worst_var,
            n_warmup = iter_warmup,
            facet_args = list(ncol = 2)
        ) + 
        theme_bw()
        
        p <- patchwork::wrap_plots(p_lp, p_other, ncol = 1, heights = c(1, 4))
        
        ggsave(
            file = paste0(output_file_prefix, "_worsttrace.png"),
            plot = p,
            h = 10,
            w = 20
        )

        # Make intervals plot
        cat("Generating parameter plots...\n")
        po <- pcm_fit$draws(
            variables = c(
                "latent_factor_unit", "latent_factor_beta",
                "skill_thresholds", "loadings_questions_m1"
            ),
            inc_warmup = FALSE,
            format = "draws_array"
        )
        po <- po[, pcm_fit_good_chains, , drop = FALSE]

        color_scheme_set("teal")
        p <- bayesplot::mcmc_intervals(
            po,
            prob = 0.5, prob_outer = 0.95, outer_size = 1, point_size = 2
        ) + theme_bw()
        ggsave(
            file = paste0(output_file_prefix, "_intervals.pdf"),
            plot = p,
            h = 50,
            w = 8,
            limitsize = FALSE
        )

        # Make posterior predictive check
        cat("Generating posterior predictive checks...\n")
        po <- pcm_fit$draws(
            variables = c("ypred"),
            inc_warmup = FALSE,
            format = "draws_array"
        )
        po <- po[, pcm_fit_good_chains, , drop = FALSE]
        po <- posterior::as_draws_df(po)
        po <- as.data.table(po)
        po <- data.table::melt(
            po,
            id.vars = c(".draw", ".chain", ".iteration"),
            value.name = "ypred"
        )
        po[, obs_idx := as.integer(gsub("ypred\\[([0-9]+)\\]", "\\1", variable))]
        set(po, NULL, "variable", NULL)

        # Map back to original observation structure using sorted dcati_copy
        dcati_copy[, obs_idx := .I]
        po <- merge(po, dcati_copy[, .(obs_idx, item_type, oidt, y_stan, oid, item_label, time_label)], 
                    by = "obs_idx")

        pos <- po[,
            list(
                summary_value = quantile(
                    ypred,
                    prob = c(0.025, 0.25, 0.5, 0.75, 0.975)
                ),
                summary_name = c("q_lower", "iqr_lower", "median", "iqr_upper", "q_upper")
            ),
            by = c("item_type", "oidt")
        ]
        pos <- data.table::dcast(pos,
            item_type + oidt ~ summary_name,
            value.var = "summary_value"
        )

        pos <- merge(pos, unique(dcati_copy[, .(item_type, oidt, y_stan, oid, item_label, time_label)]), 
                     by = c("item_type", "oidt"))
        pos[, in_ppi := y_stan >= q_lower & y_stan <= q_upper]
        cat("Proportion in 95% PPI:", pos[, mean(in_ppi)], "\n")

        # Plot posterior predictive check
        p <- ggplot(pos, aes(x = oid, group = oid)) +
            geom_boxplot(
                aes(
                    ymin = q_lower,
                    lower = iqr_lower,
                    middle = median,
                    upper = iqr_upper,
                    ymax = q_upper
                ),
                stat = "identity"
            ) +
            geom_point(aes(y = y_stan, colour = in_ppi)) +
            facet_grid(item_label ~ time_label, scales = "free") +
            scale_x_discrete() +
            scale_y_continuous() +
            ggsci::scale_color_npg() +
            labs(
                x = "",
                y = "outcome",
                colour = "within\n95% posterior\nprediction\ninterval"
            ) +
            theme_bw() +
            theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
        ggsave(
            file = paste0(output_file_prefix, "_ppcheck.png"),
            p,
            w = 20,
            h = 20
        )

        # Compute and save ordinal Brier scores by question
        cat("Computing ordinal Brier scores by question...\n")
        
        # Extract Brier scores (now in format ordinal_brier_score[c,q])
        brier <- pcm_fit$draws(
            variables = c("ordinal_brier_score"),
            inc_warmup = FALSE,
            format = "draws_array"
        )
        brier <- brier[, pcm_fit_good_chains, , drop = FALSE]
        brier <- posterior::as_draws_df(brier)
        brier <- as.data.table(brier)
        brier <- data.table::melt(brier,
            id.vars = c(".draw", ".chain", ".iteration"),
            value.name = "brier_score"
        )
        
        # Parse indices: ordinal_brier_score[c,q]
        brier[, cat_id := as.integer(gsub(
            "ordinal_brier_score\\[([0-9]+),([0-9]+)\\]", "\\1",
            variable
        ))]
        brier[, question_id := as.integer(gsub(
            "ordinal_brier_score\\[([0-9]+),([0-9]+)\\]", "\\2",
            variable
        ))]
        set(brier, NULL, "variable", NULL)
        
        # Map cat_id back to item_type
        cat_type_map <- data.table(cat_id = 1:C, item_type = item_types)
        brier <- merge(brier, cat_type_map, by = "cat_id")
        
        # Compute median
        brier_summary <- brier[,
            list(
                median_brier_score = median(brier_score),
                mean_brier_score = mean(brier_score),
                q025_brier_score = quantile(brier_score, 0.025),
                q975_brier_score = quantile(brier_score, 0.975)
            ),
            by = c("item_type", "question_id")
        ]
        
        data.table::fwrite(
            brier_summary,
            file = paste0(output_file_prefix, "_ordered_brierscore.csv")
        )
        cat("Ordinal Brier scores saved to", paste0(output_file_prefix, "_ordered_brierscore.csv"), "\n")
    }

    # Generate probability plots
    if (with_core_analyses) {
        cat("Generating probability plots...\n")
        
        # Extract ordered_prob_by_obs (long vector format)
        # NOTE: In ncats model, ordered_prob_by_obs is a single long vector [eta_length]
        # where probabilities for each observation are stored contiguously.
        # This differs from the 2cats model which has separate matrices
        # cat1_ordered_prob_by_obs[Ncat1, Kcat1] and cat2_ordered_prob_by_obs[Ncat2, Kcat2]
        po <- pcm_fit$draws(
            variables = c("ordered_prob_by_obs"),
            inc_warmup = FALSE,
            format = "draws_array"
        )
        po <- po[, pcm_fit_good_chains, , drop = FALSE]
        po <- posterior::as_draws_df(po)
        po <- as.data.table(po)
        po <- data.table::melt(po,
            id.vars = c(".draw", ".chain", ".iteration"),
            value.name = "prob"
        )
        
        # Parse index from ordered_prob_by_obs[i]
        po[, prob_idx := as.integer(gsub(
            "ordered_prob_by_obs\\[([0-9]+)\\]", "\\1",
            variable
        ))]
        set(po, NULL, "variable", NULL)
        
        # Map prob_idx back to observation and response category
        # Need to reconstruct eta_obs_offset from stan_data
        K_vec <- stan_data$K[stan_data$cat_type]  # K for each observation
        eta_obs_offset <- cumsum(c(1, K_vec[-length(K_vec)]))  # Starting position for each obs
        
        # Create mapping table
        mapping <- data.table()
        for (obs_idx in 1:stan_data$N_total) {
            K_obs <- K_vec[obs_idx]
            offset <- eta_obs_offset[obs_idx]
            obs_mapping <- data.table(
                prob_idx = offset:(offset + K_obs - 1),
                obs_idx = obs_idx,
                y_stan = 1:K_obs
            )
            mapping <- rbind(mapping, obs_mapping)
        }
        
        # Merge with probability data
        po <- merge(po, mapping, by = "prob_idx")
        
        # Merge with observation metadata
        dcati_copy[, obs_idx := .I]
        po <- merge(po, dcati_copy[, .(obs_idx, item_type, oidt, pid, item_time_id, time, 
                                       time_label, item_label)], 
                    by = "obs_idx")
        
        # Average over observations at same time/item/response
        po <- po[,
            list(prob = mean(prob)),
            by = c(".draw", "time", "item_time_id", "y_stan", "item_type")
        ]
        
        # Compute quantiles
        pos <- po[,
            list(
                summary_value = quantile(
                    prob,
                    prob = c(0.025, 0.25, 0.5, 0.75, 0.975)
                ),
                summary_name = c("q_lower", "iqr_lower", "median", "iqr_upper", "q_upper")
            ),
            by = c("time", "item_time_id", "y_stan", "item_type")
        ]
        pos <- data.table::dcast(pos,
            time + item_time_id + y_stan + item_type ~ summary_name,
            value.var = "summary_value"
        )
        
        # Add labels
        tmp <- unique(dcati_copy[, .(time, time_label)])
        pos <- merge(pos, tmp, by = "time")
        tmp <- unique(dcati_copy[, .(item_time_id, item_label, item_type)])
        pos <- merge(pos, tmp, by = c("item_time_id", "item_type"))
        tmp <- unique(dcati_copy[, .(y, y_stan, y_label)])
        pos <- merge(pos, tmp, by = "y_stan")
        
        # Compute empirical probabilities
        tmp <- dcati_copy[,
            list(n = .N),
            by = c("time_label", "item_label", "y_label")
        ]
        tmp2 <- tmp[, list(total = sum(n)), by = c("time_label", "item_label")]
        tmp <- merge(tmp, tmp2, by = c("time_label", "item_label"))
        tmp[, p_emp := n / total]
        pos <- merge(pos,
            subset(tmp, select = -c(n, total)),
            by = c("time_label", "item_label", "y_label"),
            all.x = TRUE
        )
        set(pos, pos[, which(is.na(p_emp))], "p_emp", 0.)
        pos[, y_stan := NULL]
        
        # Merge with dit for plotting
        pos <- merge(pos, dit, by = "item_label")
        pos[, item_label_long := paste0(group_label_long, " --- ", item_label_short)]

        cat(
            "\nSaving posterior probabilities to RDS with name",
            paste0(output_file_prefix, "__posterior_probabilities.rds"),
            "...\n"
        )
        saveRDS(
            pos,
            file = paste0(output_file_prefix, "__posterior_probabilities.rds")
        )

        # Make plots
        pal <- colorRampPalette(ggsci::pal_futurama("planetexpress")(12))(
            pos[, length(unique(item_label))]
        )

        p <- ggplot(
            pos,
            aes(x = y_label, group = interaction(item_label_long, y_label))
        ) +
            geom_col(aes(fill = item_label_long, y = p_emp),
                position = position_dodge(width = 0.9, preserve = "single"),
                alpha = 0.8,
                width = 0.8
            ) +
            geom_boxplot(
                aes(
                    ymin = q_lower, lower = iqr_lower,
                    middle = median, upper = iqr_upper,
                    ymax = q_upper
                ),
                position = position_dodge(0.9, preserve = "single"),
                stat = "identity",
                alpha = 0,
                width = 0.3
            ) +
            scale_y_continuous(labels = scales::percent) +
            scale_fill_manual(values = pal) +
            facet_wrap(group_label_long ~ time_label, scales = "free_x", ncol = 2) +
            theme_bw() +
            theme(
                axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
                legend.position = "bottom"
            ) +
            labs(x = "", y = "proportion of outcomes", fill = "survey items") +
            guides(fill = guide_legend(ncol = 3))
        ggsave(
            file = paste0(output_file_prefix, "_probs_barplot_v2.png"),
            plot = p,
            h = 40,
            w = 12
        )
        ggsave(
            file = paste0(output_file_prefix, "_probs_barplot_v2.pdf"),
            plot = p,
            h = 40,
            w = 12
        )
    }

    invisible(pcm_fit)
}

File: python/fit_partial_credit_model_ncats.py
"""
Model fitting functions for Bayesian IRT analysis.

This module provides functions for fitting partical credit models
using Stan HMC via cmdstanpy.
"""

from typing import Dict, Optional, Union
import os
from pathlib import Path
import warnings
import re
import time

import pandas as pd
import numpy as np
import patsy
import matplotlib.pyplot as plt
from cmdstanpy import CmdStanModel
from plotnine import *
import arviz as az

def fit_partial_credit_model_ncats(
    dit: pd.DataFrame,
    dcati: pd.DataFrame,
    output_file_prefix: str,
    stan_file: Optional[str] = None,
    x_formula: str = "time - 1",
    x_formula_ignore_regex: Optional[str] = None,
    chains: int = 2,
    parallel_chains: int = 2,
    threads_per_chain: int = 1,
    iter_warmup: int = 500,
    iter_sampling: int = 1500,
    seed: int = 123,
    show_console: bool = False,
    resume: bool = False,
    with_core_analyses: bool = True,
    with_additional_analyses: bool = False,
) -> Dict:
    """
    Run partial credit model analysis (ncats version) on pre-processed data.
    
    This function performs Bayesian IRT analysis using the flexible ncats partial credit model
    on pre-processed data. It compiles the Stan model, runs MCMC sampling, generates convergence
    diagnostics, and optionally creates detailed diagnostic plots and posterior predictive checks.
    
    Parameters
    ----------
    dit : pd.DataFrame
        Item metadata table with item types and labels.
    dcati : pd.DataFrame
        Pre-processed data with observations for analysis. Must include
        item_type, y_stan, item_time_id, pid, and oidt columns.
    output_file_prefix : str
        Full path prefix for output files (without extension).
    stan_file : str, optional
        Path to Stan model file (.stan). If None, uses default partial_credit_model_ncats_v260413.stan
    x_formula : str, default "time - 1"
        Patsy formula string specifying predictors for the design matrix.
    x_formula_ignore_regex : str, optional
        Regular expression pattern to identify X columns to exclude from posterior predictive
        probability calculations. If None, all X columns are used for both fitted and predictive.
    chains : int, default 2
        Number of MCMC chains to run.
    parallel_chains : int, default 2
        Number of chains to run in parallel.
    threads_per_chain : int, default 1
        Number of threads to use per chain.
    iter_warmup : int, default 500
        Number of warmup iterations per chain.
    iter_sampling : int, default 1500
        Number of sampling iterations per chain.
    seed : int, default 123
        Random seed for reproducibility.
    show_console : bool, default False
        If True, show Stan console output during sampling.
    resume : bool, default False
        If True and all output files exist, skip MCMC sampling and load existing results.
    with_core_analyses : bool, default True
        If True, generate core probability plots.
    with_additional_analyses : bool, default False
        If True, generate additional diagnostic plots.
    
    Returns
    -------
    dict
        Dictionary containing:
        - 'fit': CmdStanMCMC object (if not resumed)
        - 'draws': InferenceData object with posterior draws
        - 'timing': DataFrame with chain timing information
        - 'good_chains': List of chain IDs that converged
    """
    # Set default stan_file if not provided
    if stan_file is None:
        # Assume we're running from the bIRTistic root directory
        stan_file = str(Path(__file__).parents[2] / "src" / "stan" / "partial_credit_model_ncats_v260413.stan")
    
    # Print configuration
    print("\n" + "=" * 40)
    print("Partial Credit Model (ncats) Analysis Configuration")
    print("=" * 40)
    print(f"Stan file: {stan_file}")
    print(f"Stan include dir: {os.path.dirname(stan_file)}")
    print(f"Data: dit with {len(dit)} items, dcati with {len(dcati)} observations")
    print(f"Output prefix: {output_file_prefix}")
    print(f"Chains: {chains}")
    print(f"Parallel chains: {parallel_chains}")
    print(f"Threads per chain: {threads_per_chain}")
    print(f"Warmup iterations: {iter_warmup}")
    print(f"Sampling iterations: {iter_sampling}")
    print(f"Seed: {seed}")
    print(f"Resume: {resume}")
    print(f"Core analyses: {with_core_analyses}")
    print(f"Additional analyses: {with_additional_analyses}")
    print("=" * 40 + "\n")
    
    # Create output directory if it doesn't exist
    os.makedirs(os.path.dirname(output_file_prefix), exist_ok=True)
    
    # Define output file paths
    timing_file = f"{output_file_prefix}_timing.csv"
    draws_file = f"{output_file_prefix}_draws.zarr"  # Use zarr for arviz (9x faster than NetCDF)
    output_file = f"{output_file_prefix}_stan.pkl"
    mixing_file = f"{output_file_prefix}_convergence_mixing.csv"
    
    # Check if we should resume from existing outputs
    can_resume = (resume and 
                  os.path.exists(timing_file) and 
                  os.path.exists(draws_file) and 
                  os.path.exists(output_file) and 
                  os.path.exists(mixing_file))
    
    # Prepare data in ncats format
    print("Preparing Stan data in ncats format...")
    
    # Ensure dcati is sorted by item_type_id and oidt
    dcati_sorted = dcati.sort_values(['item_type_id', 'oidt']).reset_index(drop=True)
    
    # Verify oid is sequential
    assert (dcati_sorted['oid'] == dcati_sorted.index + 1).all(), \
        "oid must be sequential 1:N after sorting by item_type_id, oidt"
    
    # Build stan_data dictionary
    stan_data = {}
    stan_data['C'] = int(dcati['item_type_id'].max())
    stan_data['U'] = int(dcati['pid'].max())
    
    # Arrays for each category type
    category_stats = dcati.groupby('item_type_id').agg({
        'oid': 'count',
        'item_time_id': 'max',
        'y_stan': lambda x: len(x.unique())
    }).rename(columns={'oid': 'N', 'item_time_id': 'Q', 'y_stan': 'K'})
    
    stan_data['N'] = category_stats['N'].astype(int).tolist()
    stan_data['Q'] = category_stats['Q'].astype(int).tolist()
    stan_data['K'] = category_stats['K'].astype(int).tolist()
    
    # Total counts
    stan_data['N_total'] = int(sum(stan_data['N']))
    stan_data['Q_total'] = int(sum(stan_data['Q']))
    
    # Concatenated observation data (data is already sorted by item_type_id) - convert to int
    stan_data['y'] = dcati['y_stan'].astype(int).tolist()
    stan_data['unit_of_obs'] = dcati['pid'].astype(int).tolist()
    stan_data['question_of_obs'] = dcati['item_time_id'].astype(int).tolist()
    stan_data['cat_type'] = dcati['item_type_id'].astype(int).tolist()
    
    # Create design matrix using patsy
    design_matrix = patsy.dmatrix(x_formula, data=dcati, return_type='dataframe')
    
    # Determine which columns to use for predictive probabilities
    if x_formula_ignore_regex is not None:
        xpr_id = [i + 1 for i, col in enumerate(design_matrix.columns) 
                  if not re.search(x_formula_ignore_regex, col)]
    else:
        xpr_id = list(range(1, len(design_matrix.columns) + 1))
    
    stan_data['X'] = design_matrix.values.tolist()
    stan_data['Xpr_id'] = xpr_id
    stan_data['P'] = len(design_matrix.columns)
    stan_data['P_pr'] = len(xpr_id)
    
    print(f"Design matrix number of predictors: P = {stan_data['P']}")
    print(f"Design matrix column names (for fitting): {', '.join(design_matrix.columns)}")
    print(f"Design matrix column names (for prediction): {', '.join([design_matrix.columns[i-1] for i in xpr_id])}")
    print(f"Number of category types: C = {stan_data['C']}")
    print(f"Category type labels: {', '.join(dit['item_type'].unique())}")
    print(f"N per category: {', '.join(map(str, stan_data['N']))}")
    print(f"Q per category: {', '.join(map(str, stan_data['Q']))}")
    print(f"K per category: {', '.join(map(str, stan_data['K']))}")
    
    # Check if resuming from existing outputs
    if can_resume:
        print("\n" + "=" * 40)
        print("RESUMING from existing outputs")
        print("=" * 40)
        
        print(f"Loading draws from: {draws_file}")
        idata = az.from_zarr(draws_file)
        
        print(f"Loading timing data from: {timing_file}")
        timing_data = pd.read_csv(timing_file)
        good_chains = timing_data['chain'].tolist()
        print(f"Identified good HMC chains: {', '.join(map(str, good_chains))}")
        print("=" * 40 + "\n")
        
        return {
            'draws': idata,
            'timing': timing_data,
            'good_chains': good_chains
        }
    else:
        if resume:
            print("\nNote: Resume requested but not all output files exist. Running full analysis.\n")
        
        # Compile Stan model
        print("Compiling Stan model...")
        model = CmdStanModel(
            stan_file=stan_file,
            cpp_options={'STAN_THREADS': 'TRUE'}
        )
        
        # Sample from the model
        print("Running MCMC sampling...")
        fit = model.sample(
            data=stan_data,
            chains=chains,
            parallel_chains=parallel_chains,
            threads_per_chain=threads_per_chain,
            iter_warmup=iter_warmup,
            iter_sampling=iter_sampling,
            seed=seed,
            show_console=show_console,
            refresh=500,
            save_warmup=True
        )
        
        # Remove any chains that did not converge based on log probability
        print("Checking for divergent transitions and removing non-converged chains...")
        lp_df = fit.draws_pd(vars=['lp__'], inc_warmup=False)
        
        # Calculate mean and std per chain
        chain_stats = lp_df.groupby('chain__')['lp__'].agg(['mean', 'std', 'count']).reset_index()
        chain_stats.columns = ['chain', 'mean_lp', 'sd_lp', 'n']
        
        # Identify good chains (within 2 SD of best chain)
        best_idx = chain_stats['mean_lp'].idxmax()
        threshold = chain_stats.loc[best_idx, 'mean_lp'] - 2 * chain_stats.loc[best_idx, 'sd_lp']
        good_chains = chain_stats[chain_stats['mean_lp'] > threshold]['chain'].tolist()
        print(f"Identified good HMC chains: {', '.join(map(str, good_chains))}")
        
        # Extract chain timing information
        print("\nExtracting timing information...")
        timing_info = fit.metadata.cmdstan_config
        timing_data = pd.DataFrame({
            'chain': good_chains,
            'warmup_minutes': [timing_info['metric'] for _ in good_chains],  # Placeholder
            'sampling_minutes': [0.0 for _ in good_chains],  # Placeholder
            'total_chain_minutes': [0.0 for _ in good_chains]  # Placeholder
        })
        
        timing_data.to_csv(timing_file, index=False)
        print(f"Saved timing information to: {timing_file}")
        
        # Convert to ArviZ InferenceData
        print("Converting to ArviZ format...")
        idata = az.from_cmdstanpy(
            fit,
            coords={'chain': good_chains},
            dims={}
        )
        
        # Save draws
        print(f"Saving draws to: {draws_file}")
        idata.to_zarr(draws_file)
        
        # Check convergence and mixing
        print("Generating convergence diagnostics...")
        summary = az.summary(
            idata,
            var_names=['latent_factor_unit', 'latent_factor_beta', 
                      'skill_thresholds', 'loadings_questions_m1']
        )
        summary = summary.sort_values('ess_bulk')
        summary.to_csv(mixing_file)
        
        # Generate trace plots for worst parameters
        if with_additional_analyses:
            print("Generating trace plots...")
            worst_vars = summary.index[:9].tolist()
            
            fig, axes = plt.subplots(len(worst_vars) + 1, 1, figsize=(20, 10))
            
            # Plot log probability
            az.plot_trace(idata, var_names=['lp'], axes=axes[0:1])
            
            # Plot worst parameters
            for i, var in enumerate(worst_vars):
                az.plot_trace(idata, var_names=[var], axes=axes[i+1:i+2])
            
            plt.tight_layout()
            plt.savefig(f"{output_file_prefix}_worsttrace.png", dpi=150)
            plt.close()
        
        return {
            'fit': fit,
            'draws': idata,
            'timing': timing_data,
            'good_chains': good_chains
        }